### ISSUE TITLE
BLE issue - RPI4B 8GB HASSOS

### ISSUE THREAD

**Issue by Piffis** [icanos/hassio-plejd#176]
I'm running a RPI 4B with a fresh installation of HASSOS on a SSD (Surpervisor).

I've installed; File Editor, Mosquitto broker, Samba Share and Plejd add-ons
In Integrations I have; Home kit controller, Metrologisk Institutt, Mobile App, MQTT (Mosquitto) Raspberry Pi Power supply checker and Sonos. 

I used to run a Debian 64bit Ubuntu installation with HA supervisor manually installed and the Plejd add-on worked fine in my home. But since going to HASSOS I have this problem; 

-----------------------------------------------------------
 Add-on: Plejd
 Adds support for the Swedish home automation devices from Plejd.
-----------------------------------------------------------
 Add-on version: 0.7.0
 You are running the latest version of this add-on.
parse error: Expected string key before ':' at line 1, column 4
[10:11:27] ERROR: Unknown HTTP error occured
 System:   (aarch64 / raspberrypi4-64)
 Home Assistant Core: 2021.3.4
 Home Assistant Supervisor: 2021.03.6
-----------------------------------------------------------
 Please, share the above information when looking for help
 or support in, e.g., GitHub, forums or the Discord chat.
-----------------------------------------------------------
[cont-init.d] 00-banner.sh: exited 0.
[cont-init.d] 01-log-level.sh: executing... 
[cont-init.d] 01-log-level.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[10:11:28] INFO: Starting the Plejd service...
[10:11:28] INFO: Verified permissions on startup script
[10:11:28] INFO: Executing startup script
[10:11:28] INFO: Running add-on
Config: {
  site: 'Hemma',
  username: '---scrubbed---',
  password: '---scrubbed---',
  mqttBroker: '192.168.0.160:1883',
  mqttUsername: 'mqtt',
  mqttPassword: '---scrubbed---',
  includeRoomsAsLights: false,
  preferCachedApiResponse: false,
  updatePlejdClock: true,
  logLevel: 'info',
  connectionTimeout: 15,
  writeQueueWaitTime: 400
}
2021-03-25 10:11:29 INF [plejd-main] Log level set to info
Starting Plejd addon and reading configuration...
2021-03-25 10:11:29 INF [plejd-main] Log level set to info
2021-03-25 10:11:29 INF [plejd-main] Plejd add-on, version 0.7.0
2021-03-25 10:11:29 INF [device-comm] Starting Plejd communication handler.
2021-03-25 10:11:29 INF [plejd-ble] Starting Plejd BLE Handler, resetting all device states.
2021-03-25 10:11:29 INF [plejd-main] Main Plejd addon init()...
2021-03-25 10:11:29 INF [plejd-api] init()
2021-03-25 10:11:29 INF [plejd-api] Getting cached api response from disk
2021-03-25 10:11:29 INF [plejd-api] login()
2021-03-25 10:11:29 INF [plejd-api] logging into Hemma
2021-03-25 10:11:30 INF [plejd-api] got session token response
2021-03-25 10:11:30 INF [plejd-api] Get all Plejd sites for account...
2021-03-25 10:11:30 INF [plejd-api] Got site list response with 1: Hemma
2021-03-25 10:11:30 INF [plejd-api] Site found matching configuration name Hemma
2021-03-25 10:11:30 INF [plejd-api] Get site details for 9b72bc68-3988-4e8c-8aca-dcfca71a3805...
2021-03-25 10:11:31 INF [plejd-api] got site details response
2021-03-25 10:11:31 INF [plejd-api] Site details for site id 9b72bc68-3988-4e8c-8aca-dcfca71a3805 found
2021-03-25 10:11:31 INF [plejd-api] Saving cached copy
2021-03-25 10:11:31 INF [plejd-api] Getting devices from site details response...
2021-03-25 10:11:31 INF [plejd-mqtt] Initializing MQTT connection for Plejd addon
2021-03-25 10:11:31 INF [plejd-ble] init()
2021-03-25 10:11:36 INF [plejd-ble] BLE init done, waiting for devices.
2021-03-25 10:11:36 INF [plejd-main] Main init done
2021-03-25 10:11:36 INF [plejd-main] main() finished
2021-03-25 10:11:51 **ERR** [plejd-ble] Discovery timeout elapsed, no devices found. Starting reconnect loop...
2021-03-25 10:11:51 INF [plejd-ble] Starting reconnect loop...
2021-03-25 10:11:51 INF [device-comm] Bluetooth reconnecting...
2021-03-25 10:11:56 INF [plejd-ble] Reconnecting BLE...
2021-03-25 10:11:56 INF [plejd-ble] init()
2021-03-25 10:12:01 **ERR** [plejd-ble] Failed to start discovery. Operation already in progress
DBusError: Operation already in progress
    at _methodReturnHandlers.<computed> (/plejd/node_modules/dbus-next/lib/bus.js:343:27)
    at handleMessage (/plejd/node_modules/dbus-next/lib/bus.js:101:11)
    at EventEmitter.<anonymous> (/plejd/node_modules/dbus-next/lib/bus.js:151:9)
    at EventEmitter.emit (events.js:314:20)
    at /plejd/node_modules/dbus-next/lib/connection.js:116:14
    at Socket.<anonymous> (/plejd/node_modules/dbus-next/lib/message.js:63:9)
    at Socket.emit (events.js:314:20)
    at emitReadable_ (_stream_readable.js:557:12)
    at processTicksAndRejections (internal/process/task_queues.js:83:21)
2021-03-25 10:12:01 INF [plejd-ble] If you continue to get "operation already in progress" error, you can try power cycling the bluetooth adapter. Get root console access, run "bluetoothctl" => "power off" => "power on" => "exit" => restart addon.
2021-03-25 10:12:01 WRN [plejd-ble] Failed reconnecting. Failed to start discovery. Make sure no other add-on is currently scanning.
Error: Failed to start discovery. Make sure no other add-on is currently scanning.
    at PlejBLEHandler._startGetPlejdDevice (/plejd/PlejdBLEHandler.js:428:13)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
2021-03-25 10:12:01 INF [device-comm] Bluetooth reconnecting...
2021-03-25 10:12:06 INF [plejd-ble] Reconnecting BLE...
2021-03-25 10:12:06 INF [plejd-ble] init()

And it just loops like this... 

2021-03-25 10:12:11 **ERR** [plejd-ble] Failed to start discovery. Operation already in progress
DBusError: Operation already in progress
    at _methodReturnHandlers.<computed> (/plejd/node_modules/dbus-next/lib/bus.js:343:27)
    at handleMessage (/plejd/node_modules/dbus-next/lib/bus.js:101:11)
    at EventEmitter.<anonymous> (/plejd/node_modules/dbus-next/lib/bus.js:151:9)
    at EventEmitter.emit (events.js:314:20)
    at /plejd/node_modules/dbus-next/lib/connection.js:116:14
    at Socket.<anonymous> (/plejd/node_modules/dbus-next/lib/message.js:63:9)
    at Socket.emit (events.js:314:20)
    at emitReadable_ (_stream_readable.js:557:12)
    at processTicksAndRejections (internal/process/task_queues.js:83:21)
2021-03-25 10:12:11 INF [plejd-ble] If you continue to get "operation already in progress" error, you can try power cycling the bluetooth adapter. Get root console access, run "bluetoothctl" => "power off" => "power on" => "exit" => restart addon.
2021-03-25 10:12:11 WRN [plejd-ble] Failed reconnecting. Failed to start discovery. Make sure no other add-on is currently scanning.
Error: Failed to start discovery. Make sure no other add-on is currently scanning.
    at PlejBLEHandler._startGetPlejdDevice (/plejd/PlejdBLEHandler.js:428:13)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
2021-03-25 10:12:11 INF [device-comm] Bluetooth reconnecting...
2021-03-25 10:12:16 INF [plejd-ble] Reconnecting BLE...
2021-03-25 10:12:17 INF [plejd-ble] init()
2021-03-25 10:12:22 **ERR** [plejd-ble] Failed to start discovery. Operation already in progress
DBusError: Operation already in progress
    at _methodReturnHandlers.<computed> (/plejd/node_modules/dbus-next/lib/bus.js:343:27)
    at handleMessage (/plejd/node_modules/dbus-next/lib/bus.js:101:11)
    at EventEmitter.<anonymous> (/plejd/node_modules/dbus-next/lib/bus.js:151:9)
    at EventEmitter.emit (events.js:314:20)
    at /plejd/node_modules/dbus-next/lib/connection.js:116:14
    at Socket.<anonymous> (/plejd/node_modules/dbus-next/lib/message.js:63:9)
    at Socket.emit (events.js:314:20)
    at emitReadable_ (_stream_readable.js:557:12)
    at processTicksAndRejections (internal/process/task_queues.js:83:21)
2021-03-25 10:12:22 INF [plejd-ble] If you continue to get "operation already in progress" error, you can try power cycling the bluetooth adapter. Get root console access, run "bluetoothctl" => "power off" => "power on" => "exit" => restart addon.
2021-03-25 10:12:22 WRN [plejd-ble] Failed reconnecting. Failed to start discovery. Make sure no other add-on is currently scanning.
Error: Failed to start discovery. Make sure no other add-on is currently scanning.
    at PlejBLEHandler._startGetPlejdDevice (/plejd/PlejdBLEHandler.js:428:13)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
2021-03-25 10:12:22 INF [device-comm] Bluetooth reconnecting...

What may be the issue? As I'm running HASSOS I can't "Get root console access, run "bluetoothctl" => "power off" => "power on" => "exit" => restart addon".
It looks like the BLW is busy with something. 
I've restarted the RPI switching the power, also I can't find anyhting related to BLE in the config file available in the ROOT. 

Please help

**Comment by SweVictor** [comment]
Not sure I can help, but starting with the 0.7.0 version the addon will actually reboot the BLE adapter itself after many (100 maybe?) reconnect attempts. 

Things you could try is increase the connect timeout and move closer to a Plejd device. Turning on verbose logging might give more details, but probably not in this case.

Regarding HassOS it sounds strange you can't get root access, but I don't use it so someone else has to guide here. If you can install addons maybe you can use the SSH/Terminal addon which has unsecured root access option?

**Comment by Piffis** [comment]
Yes, it reboots after 70, 110 attempts (and still going :) ) but I have the same issue still. 

I've increased the timeout to 25, it used to be ok with 10 when I ran HA supervised on Ubuntu. 

I'm no HassOS expert but from what I understand it's a "closed" OS. Ie I can't SSH into the root. I can install the SSH add on of course. But after that? It won't respond to any ssh user@"ip of machine" calls. I tried typing bluetoothctrl just in case the SSH add on actually took me to the root by default.. But it didn't seem to do anything. I'm not very good in Linux so I'm not sure how to type the commands:
"bluetoothctl" => "power off" => "power on" => "exit" => restart addon"

See my attempts in the picture attached
![SSH](https://user-images.githubusercontent.com/81358293/112518715-ed8ebf00-8d99-11eb-9da2-15d7ed90efe8.JPG)

Do I understand the log correct that the built in Bluetooth adaptor is busy somehow? Ie it's not the Plejd device that isn't responding, it's the Bluetooth adapter itself?

**Comment by faanskit** [comment]
> I can't SSH into the root

Well, you can. [This](https://developers.home-assistant.io/docs/operating-system/debugging/) page tells you how to do it:

1. Use a USB drive formatted with FAT, ext4, or NTFS and name it **CONFIG** (case sensitive). 
2. Create an **authorized_keys file** (no extension) containing your public key, and place it in the root of the USB drive. File needs to be ANSI encoded (not UTF-8) and must have Unix line ends (LF), not Windows (CR LF). 
3. From the UI, navigate to the Supervisor system page and choose "Import from USB". 
4. You can now access your device as root over SSH on port 22222. Alternatively, the file will be imported from the USB when the Home Assistant OS device is rebooted.

If you already have a private key, load it into puttygen.exe and there you will find the data to put into authorized_keys. Remember to configure putty to use your private key (private.ppk). Connection/SSH/Auth/Private key for authentification

Connect to port 22222 via SSH and login as root.
Now you will be in a HA environment. To enter the OS, simply enter `login`

If you need grab longer logs than in the Ui, follow the instructions in the [logs](https://github.com/icanos/hassio-plejd/tree/master/plejd#logs) section.

**Comment by sundberg84** [comment]
I have the same issue, worked great before on my RPi but SD card crashed, so new fresh install (now on ssd) HassOS, Plejd and Mosquito - nothing else. Logged into root and device are ok i think: 

`~ $ bluetoothctl
Agent registered
[CHG] Controller B8:..:3D Pairable: yes
[bluetooth]# list
Controller B8:...:3D homeassistant [default]
[bluetooth]# devices
Device E3:..:FA P mesh`

Logs:
`2021-04-30 08:42:09 INF [plejd-main] Log level set to info
Starting Plejd addon and reading configuration...
2021-04-30 08:42:09 INF [plejd-main] Log level set to info
2021-04-30 08:42:09 INF [plejd-main] Plejd add-on, version 0.7.1
2021-04-30 08:42:09 INF [device-comm] Starting Plejd communication handler.
2021-04-30 08:42:09 INF [plejd-ble] Starting Plejd BLE Handler, resetting all device states.
2021-04-30 08:42:09 INF [plejd-main] Main Plejd addon init()...
2021-04-30 08:42:09 INF [plejd-api] init()
2021-04-30 08:42:09 INF [plejd-api] Getting cached api response from disk
2021-04-30 08:42:09 INF [plejd-api] login()
2021-04-30 08:42:09 INF [plejd-api] logging into SundbergHome
2021-04-30 08:42:10 INF [plejd-api] got session token response
2021-04-30 08:42:10 INF [plejd-api] Get all Plejd sites for account...
2021-04-30 08:42:10 INF [plejd-api] Got site list response with 1: SundbergHome
2021-04-30 08:42:10 INF [plejd-api] Site found matching configuration name SundbergHome
2021-04-30 08:42:10 INF [plejd-api] Get site details for 6615cb4c-5be9-49f1-9d7e-2821902f2f92...
2021-04-30 08:42:11 INF [plejd-api] got site details response
2021-04-30 08:42:11 INF [plejd-api] Site details for site id 6615cb4c-5be9-49f1-9d7e-2821902f2f92 found
2021-04-30 08:42:11 INF [plejd-api] Saving cached copy
2021-04-30 08:42:11 INF [plejd-api] Getting devices from site details response...
2021-04-30 08:42:11 INF [plejd-mqtt] Initializing MQTT connection for Plejd addon
2021-04-30 08:42:12 INF [plejd-ble] init()
2021-04-30 08:42:13 INF [plejd-mqtt] Connected to MQTT.
2021-04-30 08:42:14 INF [plejd-mqtt] Discovered light (DIM-01) named Köksbel with PID 11.
2021-04-30 08:42:14 INF [plejd-mqtt] Discovered light (DIM-01) named Allrum downlight  with PID 13.
2021-04-30 08:42:14 INF [plejd-mqtt] Discovered light (DIM-01) named Korridor with PID 15.
2021-04-30 08:42:14 INF [plejd-mqtt] Discovered light (DIM-01) named Downlights bortre 2 with PID 16.
2021-04-30 08:42:14 INF [plejd-mqtt] Discovered light (LED-10) named Minispottar kök with PID 19.
2021-04-30 08:42:14 INF [plejd-mqtt] Discovered light (LED-10) named Minispottar höger with PID 20.
2021-04-30 08:42:14 INF [plejd-mqtt] Discovered light (LED-10) named Minispottar vänster allrum with PID 21.
2021-04-30 08:42:14 INF [plejd-mqtt] Discovered light (CTR-01) named Ytterbelysning  with PID 23.
2021-04-30 08:42:14 INF [plejd-mqtt] Discovered switch (Scene) named Släck allt with PID 1.
2021-04-30 08:42:17 ERR [plejd-ble] Error handling /org/bluez/hci0/dev_E3_E3_DB_8B_A2_FA Resource Not Ready
DBusError: Resource Not Ready
    at _methodReturnHandlers.<computed> (/plejd/node_modules/dbus-next/lib/bus.js:343:27)
    at handleMessage (/plejd/node_modules/dbus-next/lib/bus.js:101:11)
    at EventEmitter.<anonymous> (/plejd/node_modules/dbus-next/lib/bus.js:151:9)
    at EventEmitter.emit (events.js:314:20)
    at /plejd/node_modules/dbus-next/lib/connection.js:116:14
    at Socket.<anonymous> (/plejd/node_modules/dbus-next/lib/message.js:63:9)
    at Socket.emit (events.js:314:20)
    at emitReadable_ (_stream_readable.js:557:12)
    at processTicksAndRejections (internal/process/task_queues.js:83:21)
(node:302) UnhandledPromiseRejectionWarning: DBusError: Resource Not Ready
    at _methodReturnHandlers.<computed> (/plejd/node_modules/dbus-next/lib/bus.js:343:27)
    at handleMessage (/plejd/node_modules/dbus-next/lib/bus.js:101:11)
    at EventEmitter.<anonymous> (/plejd/node_modules/dbus-next/lib/bus.js:151:9)
    at EventEmitter.emit (events.js:314:20)
    at /plejd/node_modules/dbus-next/lib/connection.js:116:14
    at Socket.<anonymous> (/plejd/node_modules/dbus-next/lib/message.js:63:9)
    at Socket.emit (events.js:314:20)
    at emitReadable_ (_stream_readable.js:557:12)
    at processTicksAndRejections (internal/process/task_queues.js:83:21)
(node:302) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:302) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
2021-04-30 08:42:17 ERR [plejd-ble] Failed to start discovery. Resource Not Ready
DBusError: Resource Not Ready
    at _methodReturnHandlers.<computed> (/plejd/node_modules/dbus-next/lib/bus.js:343:27)
    at handleMessage (/plejd/node_modules/dbus-next/lib/bus.js:101:11)
    at EventEmitter.<anonymous> (/plejd/node_modules/dbus-next/lib/bus.js:151:9)
    at EventEmitter.emit (events.js:314:20)
    at /plejd/node_modules/dbus-next/lib/connection.js:116:14
    at Socket.<anonymous> (/plejd/node_modules/dbus-next/lib/message.js:63:9)
    at Socket.emit (events.js:314:20)
    at emitReadable_ (_stream_readable.js:557:12)
    at processTicksAndRejections (internal/process/task_queues.js:83:21)
2021-04-30 08:42:17 ERR [device-comm] Failed init() of BLE. Starting reconnect loop.
2021-04-30 08:42:17 INF [plejd-ble] Starting reconnect loop...
2021-04-30 08:42:17 INF [device-comm] Bluetooth reconnecting...`

**Comment by SweVictor** [comment]
What happened with this? Any progress? Not really sure how to help fixing this unfortunately.

**Comment by sundberg84** [comment]
> What happened with this? Any progress? Not really sure how to help fixing this unfortunately.

I tried downgrading to older version which worked before, but same issue.
I have ordered an external USB bluetooth dongle to try that. While debugging and watching the log for bluetoothctl it seems like the devices is found in the discovery, and first add them and then removed them again. Then everything hangs from there and only thing to do is to power on/off the device.
